packages/dds/tree/src/codec/codec.ts

```diff
@@ -0,0 +1,177 @@
+ * the `formatVersion` parameter)
+ * allows avoiding some duplicate work at encode/decode time, since the vast majority of document usage will not
+ * involve mixed format versions.
+ * @alpha
+ */
+export interface ICodecFamily<TDecoded> {
+	/**
+	 * @returns - a codec that can be used to encode and decode data in the specified format.
+	 */
+	resolve(formatVersion: number): IMultiFormatCodec<TDecoded>;
+
+	/**
+	 * @returns - an iterable of all format versions supported by this family.
+	 */
+	getSupportedFormats(): Iterable<number>;
+}
+
+/**
+ * Creates a codec family from a registry of codecs.
+ * Any codec that is not a {@link IMultiFormatCodec} will be wrapped with a default binary encoding.
+ */
+export function makeCodecFamily<TDecoded>(
+	registry: Iterable<
+		[formatVersion: number, codec: IMultiFormatCodec<TDecoded> | IJsonCodec<TDecoded>]
+	>,
+): ICodecFamily<TDecoded> {
+	const isJsonCodec = (
+		codec: IMultiFormatCodec<TDecoded> | IJsonCodec<TDecoded>,
+	): codec is IJsonCodec<TDecoded> =>
+		typeof (codec as any).encode === "function" && typeof (codec as any).decode === "function";
+	const codecs: Map<number, IMultiFormatCodec<TDecoded>> = new Map();
+	for (const [formatVersion, codec] of registry) {
+		if (codecs.get(formatVersion) !== undefined) {
+			fail("Duplicate codecs specified.");
+		}
+		if (isJsonCodec(codec)) {
+			codecs.set(formatVersion, withDefaultBinaryEncoding(codec));
+		} else {
+			codecs.set(formatVersion, codec);
+		}
+	}
+
+	return {
+		resolve(formatVersion: number) {
+			const codec = codecs.get(formatVersion);
+			assert(codec !== undefined, "Requested coded for unsupported format.");
+			return codec;
+		},
+		getSupportedFormats() {
+			return codecs.keys();
+		},
+	};
+}
+
+class DefaultBinaryCodec<TDecoded> implements IBinaryCodec<TDecoded> {
+	public constructor(private readonly jsonCodec: IJsonCodec<TDecoded>) {}
+
+	public encode(change: TDecoded): IsoBuffer {
+		const jsonable = this.jsonCodec.encode(change);
+		const json = JSON.stringify(jsonable);
+		return IsoBuffer.from(json);
+	}
+
+	public decode(change: IsoBuffer): TDecoded {
+		const json = bufferToString(change, "utf8");
+		const jsonable = JSON.parse(json);
+		return this.jsonCodec.decode(jsonable);
+	}
+}
+
+/**
+ * Constructs a {@link IMultiFormatCodec} from a `IJsonCodec` using a generic binary encoding that simply writes
+ * the json representation of the object to a buffer.
+ */
+export function withDefaultBinaryEncoding<TDecoded>(
+	jsonCodec: IJsonCodec<TDecoded>,
+): IMultiFormatCodec<TDecoded> {
+	return {
+		json: jsonCodec,
+		binary: new DefaultBinaryCodec(jsonCodec),
+	};
+}
+
+/**
+ * Codec for objects which carry no information.
+ */
+export const unitCodec: IMultiFormatCodec<0> = {
+	json: {
+		encode: () => 0,
+		decode: () => 0,
+	},
+	binary: {
+		encode: () => IsoBuffer.from(""),
+		decode: () => 0,
+	},
+};
+
+/**
+ * Creates a json codec for objects which are just a json compatible value
```

Comment:
Ins't this an anti-pattern? We want to strongly encourage people not to take an existing type and use it for persistence, since then someone might modify that type and end up causing a data compat issue. At the very leasy this deserves a massive disclaimer that the provided type needs to be one that never changes.

Taking a TypeBox schema instead would help emphasize that.

Reply:
I agree this isn't a good pattern but it's pre-existing/out of scope of this PR. I'm guessing current code mostly uses this out of convenience. I think the example you provide in the first paragraph is perhaps the less dangerous part of this pattern: if you happen to use in-memory types that align with your serialized format, but the type is defined in some sort of format file, it should be pretty obvious you can't change it inline without compatibility issues.

The part that scares me more about this pattern is the possibility of transparently assigning extra data to one of these types at runtime, and ending up either including large amounts of data accidentally at serialization time or serialization failing entirely (ex: you put circular references). I like the TypeBox schema suggestion for that reason, with some documentation that you should generally not allow extra properties on types used with one of these (else you get the same problem).

I'll make a note of this here and file some follow-up work.